### PR TITLE
[Feature] Show sub-process task info in workflow definition page

### LINKED ISSUE #10546: [Feature] Show sub-process task info in workflow definition page.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

I have defined several sub-workflows in Workflow, but their tasks are not displayed, so they have to jump to see them. Can we support direct expansion, so that we can see the situation of workflow intuitively. I can support this feature if it can do.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by stalary**
## Brief change log

截屏2022-06-23 上午12 20 03
截屏2022-06-23 上午12 19 53

## Related issues
close https://github.com/apache/dolphinscheduler/issues/10546

### CODE REVIEW

**Comment by stalary** [comment]
At present, I have not achieved the location well, I wonder if there is any good way.

**Comment by stalary** [comment]
I have a few questions.
1. do we recursively show all sub-process tasks (there may be multiple sub-process tasks)
2. how do we relate tasks under subtasks of the current task
3. How to determine location

**Comment by stalary** [comment]
I am not familiar with the frontend. At present, I reserve the showSubTask param for the frontend, and the frontend needs to add a button, which will trigger the process-definition interface of showSubTask=true when clicked and format the dag.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionController.java [hunk 1]
```diff
@@ -398,8 +398,9 @@ public Result releaseProcessDefinition(@ApiIgnore @RequestAttribute(value = Cons
     @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
     public Result queryProcessDefinitionByCode(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
                                                @ApiParam(name = "projectCode", value = "PROJECT_CODE", required = true) @PathVariable long projectCode,
-                                               @PathVariable(value = "code", required = true) long code) {
-        Map<String, Object> result = processDefinitionService.queryProcessDefinitionByCode(loginUser, projectCode, code);
+                                               @PathVariable(value = "code", required = true) long code,
+                                               @RequestParam(required = false) boolean showSubTask) {
+        Map<String, Object> result = processDefinitionService.queryProcessDefinitionByCode(loginUser, projectCode, code, showSubTask);
```
Comment (SbloodyS) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionController.java:408]: Please add this param to the swagger's doc `@ApiImplicitParam`.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java [hunk 2]
```diff
@@ -27,6 +29,7 @@
 import static org.apache.dolphinscheduler.common.Constants.CMD_PARAM_SUB_PROCESS_DEFINE_CODE;
 import static org.apache.dolphinscheduler.common.Constants.CMD_PARAM_SUB_PROCESS_PARENT_INSTANCE_ID;
 import static org.apache.dolphinscheduler.common.Constants.LOCAL_PARAMS;
+import org.apache.dolphinscheduler.plugin.task.api.TaskConstants;
```
Comment (SbloodyS) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]: Please format this with ds checkstyle.

**Comment by SbloodyS** [comment]
BTW. There are some unit-test failed. @stalary

**Comment by stalary** [comment]
> BTW. There are some unit-test failed. @stalary

Ok, I'll check it

**Comment by stalary** [comment]
So far I have implemented frontend click to show sub-process task, but not automatic format, and the current implementation is not so good.

**Comment by stalary** [comment]
@caishunfeng @songjianet @Amy0104 I have finished the development of the feature, PTAL.

**Comment by SbloodyS** [comment]
I had restarted the timeout CI.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java [hunk 3]
```diff
@@ -2649,17 +2653,62 @@ public DAG<String, TaskNode, TaskNodeRelation> genDagGraph(ProcessDefinition pro
...
+            if (TaskConstants.TASK_TYPE_SUB_PROCESS.equals(taskDefinition.getTaskType())) {
+                JsonNode processDefinitionCode = JSONUtils.parseObject(taskDefinition.getTaskParams())
+                    .findValue("processDefinitionCode");
+                long subProcessId = processDefinitionCode.asLong();
+                ProcessDefinition subProcess = processDefineMapper.queryByCode(subProcessId);
+                List<ProcessTaskRelation> subTaskRelations = findRelationByCode(subProcess.getCode(), subProcess.getVersion());
+                // bind to subProcess
+                for (ProcessTaskRelation subTaskRelation : subTaskRelations) {
+                    if (subTaskRelation.getPreTaskCode() == 0) {
+                        subTaskRelation.setPreTaskCode(taskDefinition.getCode());
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java:2673]: Why to set taskDefinition code and version to preTask? Please add some notes.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java [hunk 4]
```diff
@@ -2649,17 +2653,62 @@ public DAG<String, TaskNode, TaskNodeRelation> genDagGraph(ProcessDefinition pro
...
+        }
+    }
+
     /**
      * generate DagData
      */
     @Override
-    public DagData genDagData(ProcessDefinition processDefinition) {
-        List<ProcessTaskRelation> taskRelations = this.findRelationByCode(processDefinition.getCode(), processDefinition.getVersion());
+    public DagData genDagData(ProcessDefinition processDefinition, boolean showSubTask) {
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java:2700]: please update the notes of `genDagData`.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java [hunk 5]
```diff
@@ -2649,17 +2653,62 @@ public DAG<String, TaskNode, TaskNodeRelation> genDagGraph(ProcessDefinition pro
         return DagHelper.buildDagGraph(processDag);
     }
 
+    private void genSubTaskData(List<ProcessTaskRelation> taskRelations, List<TaskDefinitionLog> taskDefinitionLogList,
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]: please add notes.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java [hunk 6]
```diff
@@ -2649,17 +2653,62 @@ public DAG<String, TaskNode, TaskNodeRelation> genDagGraph(ProcessDefinition pro
...
-    public DagData genDagData(ProcessDefinition processDefinition) {
-        List<ProcessTaskRelation> taskRelations = this.findRelationByCode(processDefinition.getCode(), processDefinition.getVersion());
+    public DagData genDagData(ProcessDefinition processDefinition, boolean showSubTask) {
+        List<ProcessTaskRelation> taskRelations = findRelationByCode(processDefinition.getCode(), processDefinition.getVersion());
         List<TaskDefinitionLog> taskDefinitionLogList = genTaskDefineList(taskRelations);
-        List<TaskDefinition> taskDefinitions = taskDefinitionLogList.stream().map(t -> (TaskDefinition) t).collect(Collectors.toList());
+        List<TaskDefinition> taskDefinitions = taskDefinitionLogList.stream().map(TaskDefinition.class::cast).collect(Collectors.toList());
+        // when enable showSubTask try to find sub-process task info.
+        if (showSubTask) {
+            genSubTaskData(taskRelations, taskDefinitionLogList, taskDefinitions, taskDefinitions);
```
Comment (caishunfeng) [inline dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]: The third param is the same as the forth one, is it correct?

File: dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionControllerTest.java [hunk 7]
```diff
@@ -195,8 +195,8 @@ public void testQueryProcessDefinitionByCode() {
         putMsg(result, Status.SUCCESS);
         result.put(Constants.DATA_LIST, processDefinition);
 
-        Mockito.when(processDefinitionService.queryProcessDefinitionByCode(user, projectCode, code)).thenReturn(result);
-        Result response = processDefinitionController.queryProcessDefinitionByCode(user, projectCode, code);
+        Mockito.when(processDefinitionService.queryProcessDefinitionByCode(user, projectCode, code, false)).thenReturn(result);
```
Comment (caishunfeng) [inline dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionControllerTest.java:198]: can you add the test case of `showSubtask = true`?

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java (hunk 3 above)
Comment (stalary) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java:2673]: Use to associate a downstream child task with a parent task.

File: dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java (hunk 6 above)
Comment (stalary) [reply dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/process/ProcessServiceImpl.java]: One save all data, and the other save the current data when recursively.

File: dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionControllerTest.java (hunk 7 above)
Comment (stalary) [reply dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/controller/ProcessDefinitionControllerTest.java:198]: I met a problem in the ut, I need to query the sub-process in the database, do you know how I should generate this part of the test data.

**Review by SbloodyS — COMMENTED** [review]
Generally LGTM. Please resolve conflicts.

**Pull request closed without merging**